lib/posix/pthread_mutex.c

```diff
@@ -9,16 +9,199 @@
 #include <wait_q.h>
 #include <posix/pthread.h>
 
+#define MUTEX_MAX_REC_LOCK 32767
+/* Memory pool for pthread space */
+K_MEM_POOL_DEFINE(posix_mutex_sem, sizeof(struct k_sem),
+		  sizeof(struct k_sem), 32, 4);
+
+/*
+ *  Default mutex attrs.
+ */
+static const pthread_mutexattr_t def_attr = {
+	.type = PTHREAD_MUTEX_DEFAULT,
+};
+
+static int acquire_mutex(pthread_mutex_t *m, int timeout)
+{
+	int rc = 0, type, count;
+	pthread_t owner;
+	int key = irq_lock();
+
+	owner = m->owner;
+	type = m->type;
+	count = m->lock_count;
+	irq_unlock(key);
+
+	if (owner == pthread_self()) {
+		if (type == PTHREAD_MUTEX_ERRORCHECK) {
+			return EDEADLK;
+		}
+	}
+
+	if (type == PTHREAD_MUTEX_RECURSIVE) {
+
+		if ((owner == pthread_self()) || (count == 0)) {
+			if (count >= MUTEX_MAX_REC_LOCK) {
+				return EAGAIN;
+			}
+			rc = -k_sem_take(m->sem, timeout);
+			count++;
+		} else {
+			rc = EPERM;
+		}
+	} else {
+		rc = k_sem_take(m->sem, timeout);
+		count = 1;
+	}
+
+	if (rc == 0) {
+		key = irq_lock();
+		m->owner = pthread_self();
+		m->lock_count = count;
```

Comment:
Here is an example of the race @lpereira was talking about: you're setting count here to a value computed **outside** the critical section!  There is no guarantee that someone else didn't muck with the mutex while the irq lock was released, so the value you computed above for "count" may be wrong here.